embassy-sync/src/watch.rs

```diff
@@ -0,0 +1,1133 @@
+}
+
+impl<M: RawMutex, T: Clone, const N: usize> SealedWatchBehavior<T> for Watch<M, T, N> {
+    fn poll_get(&self, id: &mut u64, cx: &mut Context<'_>) -> Poll<T> {
+        self.mutex.lock(|state| {
+            let mut s = state.borrow_mut();
+            match &s.data {
+                Some(data) => {
+                    *id = s.current_id;
+                    Poll::Ready(data.clone())
+                }
+                None => {
+                    s.wakers.register(cx.waker());
+                    Poll::Pending
+                }
+            }
+        })
+    }
+
+    fn poll_get_and(&self, id: &mut u64, f: &mut dyn Fn(&T) -> bool, cx: &mut Context<'_>) -> Poll<T> {
+        self.mutex.lock(|state| {
+            let mut s = state.borrow_mut();
+            match s.data {
+                Some(ref data) if f(data) => {
+                    *id = s.current_id;
+                    Poll::Ready(data.clone())
+                }
+                _ => {
+                    s.wakers.register(cx.waker());
+                    Poll::Pending
+                }
+            }
+        })
+    }
+
+    fn poll_changed(&self, id: &mut u64, cx: &mut Context<'_>) -> Poll<T> {
+        self.mutex.lock(|state| {
+            let mut s = state.borrow_mut();
+            match (&s.data, s.current_id > *id) {
+                (Some(data), true) => {
+                    *id = s.current_id;
+                    Poll::Ready(data.clone())
+                }
+                _ => {
+                    s.wakers.register(cx.waker());
+                    Poll::Pending
+                }
+            }
+        })
+    }
+
+    fn try_changed(&self, id: &mut u64) -> Option<T> {
+        self.mutex.lock(|state| {
+            let s = state.borrow();
+            match s.current_id > *id {
+                true => {
+                    *id = s.current_id;
+                    s.data.clone()
+                }
+                false => None,
+            }
+        })
+    }
+
+    fn poll_changed_and(&self, id: &mut u64, f: &mut dyn Fn(&T) -> bool, cx: &mut Context<'_>) -> Poll<T> {
+        self.mutex.lock(|state| {
+            let mut s = state.borrow_mut();
+            match (&s.data, s.current_id > *id) {
+                (Some(data), true) if f(data) => {
+                    *id = s.current_id;
+                    Poll::Ready(data.clone())
+                }
+                _ => {
+                    s.wakers.register(cx.waker());
+                    Poll::Pending
+                }
+            }
+        })
+    }
+
+    fn try_changed_and(&self, id: &mut u64, f: &mut dyn Fn(&T) -> bool) -> Option<T> {
+        self.mutex.lock(|state| {
+            let s = state.borrow();
+            match (&s.data, s.current_id > *id) {
+                (Some(data), true) if f(data) => {
+                    *id = s.current_id;
+                    s.data.clone()
+                }
+                _ => None,
+            }
+        })
+    }
```

Comment:
There appears to be a lot of code duplication in these implementations. Could you combine several of these into a single internal one on the Watch type that these could use?

This applies to the different send_ and get_ variants as well.

Reply:
I can combine all the `poll, try, changed, get, and` into a single monolith:
```rust
fn receive_inner(
    &self,
    id: Option<&mut u64>,
    f: Option<&mut dyn Fn(&T) -> bool>,
    cx: Option<&mut Context<'_>>,
    must_change: bool,
) -> Poll<T> {
    self.mutex.lock(|state| {
        let mut s = state.borrow_mut();

        // If we require the value to have changed, but the state ID is not larger
        // than the receiver ID, we register the waker and return pending.
        if must_change && !id.as_ref().is_some_and(|id| s.current_id > **id) {
            cx.map(|cx| s.wakers.register(cx.waker()));
            return Poll::Pending;
        }

        // If the data does not exist, we register the waker and return pending.
        let Some(data) = &s.data else {
            cx.map(|cx| s.wakers.register(cx.waker()));
            return Poll::Pending;
        };

        // If the predicate exists, but it does not match,
        // we register the waker and return pending.
        if f.map_or(false, |f| !f(data)) {
            cx.map(|cx| s.wakers.register(cx.waker()));
            return Poll::Pending;
        }

        // Update the receiver ID, if any, and return value
        id.map(|id| *id = s.current_id);
        return Poll::Ready(data.clone());
    })
}
```
I guess this would still get optimized well, since user-code can only access this functionality through functions where the `Option`s have a known variant, like:

```rust
/// Waits for the `Watch` to change and returns the new value, marking it as seen.
pub async fn changed(&mut self) -> T {
    poll_fn(|cx| self.watch.receive_inner(Some(&mut self.at_id), None, Some(cx), true)).await
}

/// Tries to get the new value of the watch without waiting, marking it as seen.
pub fn try_changed(&mut self) -> Option<T> {
    match self.watch.receive_inner(Some(&mut self.at_id), None, None, true) {
        Poll::Ready(data) => Some(data),
        Poll::Pending => None,
    }
}
```

Reply:
Hmm, yeah. There was more nuance than I saw initially, maybe keep the original then.